src/lewis/adapters/stream.py

```diff
@@ -37,27 +37,63 @@ class StreamHandler(asynchat.async_chat):
     def __init__(self, sock, target, stream_server):
         asynchat.async_chat.__init__(self, sock=sock)
         self.set_terminator(b(target.in_terminator))
-        self.target = target
-        self.buffer = []
+        self._readtimeout = target.readtimeout
+        self._readtimer = 0
+        self._target = target
+        self._buffer = []
 
         self._stream_server = stream_server
 
         self._set_logging_context(target)
         self.log.info('Client connected from %s:%s', *sock.getpeername())
 
+    def process(self, msec):
+        if not self._buffer:
+            return
+
+        if self._readtimer >= self._readtimeout and self._readtimeout != 0:
+            if not self.get_terminator():
+                # If no terminator is set, this timeout is the terminator
+                self.found_terminator()
+            else:
+                self._readtimer = 0
+                request = self._get_request()
+                with self._stream_server.device_lock:
+                    error = RuntimeError("ReadTimeout while waiting for command terminator.")
+                    reply = self._handle_error(request, error)
+                self._send_reply(reply)
+
+        if self._buffer:
+            self._readtimer += msec
```

Comment:
After thinking about it for a bit, I moved this down to the bottom with another check against `_buffer`.

This is because, when it is at the top, we could trigger a timeout on the same cycle as we first received data (if `readtimeout <= 10 ms`. This wouldn't be good, we should wait at least one cycle. `_buffer` needs to be checked again because it might have been consumed by the code above.